File: packages/core/src/modules/secrets-scan.ts
/**
 * R6 — Secrets / Hardcoded Credentials Scan Module
 * Detects AWS keys, GitHub tokens, private key blocks
 */

import * as fs from 'fs';
import * as path from 'path';
import { ScanContext, ScanFinding, ScannerModule } from '../types';

const SECRET_PATTERNS: Array<{
  name: string;
  regex: RegExp;
  message: string;
}> = [
  {
    name: 'aws-access-key',
    regex: /AKIA[0-9A-Z]{16}/g,
    message: 'Hardcoded AWS access key detected',
  },
  {
    name: 'github-token',
    regex: /gh[pousr]_[A-Za-z0-9_]{36,}/g,
    message: 'Hardcoded GitHub token detected',
  },
  {
    name: 'private-key-block',
    regex: /-----BEGIN (RSA |OPENSSH |PGP |EC )?PRIVATE KEY-----/g,
    message: 'Private key block detected',
  },
  {
    name: 'slack-token',
    regex: /xox[baprs]-[0-9a-zA-Z\-]+/g,
    message: 'Hardcoded Slack token detected',
  },
  {
    name: 'generic-api-key',
    regex: /(?:api[_-]?key|apikey|api[_-]?secret)["']?\s*[:=]\s*["']([a-zA-Z0-9_\-]{16,})["']/gi,
    message: 'Possible hardcoded API key detected',
  },
];

export class SecretsScanModule implements ScannerModule {
  name = 'secrets-scan';

  scan(ctx: ScanContext): ScanFinding[] {
    const findings: ScanFinding[] = [];
    const codeFiles = ctx.sourceFiles.filter((f) =>
      /\.(js|ts|jsx|tsx|py|mjs|cjs|json|yaml|yml|md)$/.test(f)
    );

    for (const relPath of codeFiles) {
      const fullPath = path.join(ctx.skillPath, relPath);
      const content = fs.readFileSync(fullPath, 'utf-8');

      for (const pattern of SECRET_PATTERNS) {
        const matches = content.match(pattern.regex);
        if (matches) {
          for (const match of matches) {
            findings.push({
              ruleId: 'R6-secrets-scan',
              tier: 'blocker',
              severity: 'P0',
              criticalTag: '[critical:security]',
              message: `${pattern.message}: ${match.substring(0, 20)}...`,
              file: relPath,
              category: 'malicious-code',
              evidence: match.substring(0, 40),
              recommendation: 'Remove hardcoded secrets; use @quickport/orchestrator/credentials/* API',
            });
          }
        }
      }
    }

    return findings;
  }
}

File: crates/core/src/modules/secrets_scan.rs
use std::fs;
use std::path::Path;

use regex::Regex;

use crate::types::{
    CriticalTag, RuleOrigin, ScanContext, ScanFinding, ScannerModule, Severity, ThreatCategory,
    Tier,
};

struct SecretPattern {
    regex: Regex,
    message: &'static str,
}

fn secret_patterns() -> Vec<SecretPattern> {
    vec![
        SecretPattern {
            regex: Regex::new(r"AKIA[0-9A-Z]{16}").unwrap(),
            message: "Hardcoded AWS access key detected",
        },
        SecretPattern {
            regex: Regex::new(r"gh[pousr]_[A-Za-z0-9_]{36,}").unwrap(),
            message: "Hardcoded GitHub token detected",
        },
        SecretPattern {
            regex: Regex::new(r"-----BEGIN (?:RSA |OPENSSH |PGP |EC )?PRIVATE KEY-----").unwrap(),
            message: "Private key block detected",
        },
        SecretPattern {
            regex: Regex::new(r"xox[baprs]-[0-9a-zA-Z\-]+").unwrap(),
            message: "Hardcoded Slack token detected",
        },
        SecretPattern {
            regex: Regex::new(
                r#"(?i)(?:api[_\-]?key|apikey|api[_\-]?secret)["']?\s*[:=]\s*["']([a-zA-Z0-9_\-]{16,})["']"#,
            )
            .unwrap(),
            message: "Possible hardcoded API key detected",
        },
    ]
}

pub struct SecretsScanModule;

impl ScannerModule for SecretsScanModule {
    fn name(&self) -> &str {
        "secrets-scan"
    }

    fn scan(&self, ctx: &ScanContext) -> Vec<ScanFinding> {
        let mut findings = Vec::new();
        let code_ext_re =
            Regex::new(r"\.(js|ts|jsx|tsx|py|mjs|cjs|json|yaml|yml|md)$").unwrap();
        let code_files: Vec<&String> = ctx
            .source_files
            .iter()
            .filter(|f| code_ext_re.is_match(f))
            .collect();

        let patterns = secret_patterns();

        for rel_path in code_files {
            let full_path = Path::new(&ctx.skill_path).join(rel_path);
            let content = match fs::read_to_string(&full_path) {
                Ok(c) => c,
                Err(_) => continue,
            };

            for pattern in &patterns {
                for m in pattern.regex.find_iter(&content) {
                    let matched = m.as_str();
                    let display = if matched.len() > 20 {
                        format!("{}...", &matched[..20])
                    } else {
                        matched.to_string()
                    };
                    findings.push(ScanFinding {
                        rule_id: "R6-secrets-scan".to_string(),
                        tier: Tier::Blocker,
                        severity: Severity::P0,
                        critical_tag: Some(CriticalTag::Security),
                        message: format!("{}: {}", pattern.message, display),
                        file: Some(rel_path.clone()),
                        line: None,
                        column: None,
                        category: ThreatCategory::MaliciousCode,
                        evidence: Some(matched[..matched.len().min(40)].to_string()),
                        recommendation: Some(
                            "Remove hardcoded secrets; use @quickport/orchestrator/credentials/* API"
                                .to_string(),
                        ),
                        rule_origin: Some(RuleOrigin::Core),
                        ref_anchor: None,
                        merged_from: None,
                    });
                }
            }
        }

        findings
    }
}
